src/dscanner/analysis/vcall_in_ctor.d

```diff
@@ -4,431 +4,166 @@
-            {
-                pushVirtual(false);
-                pop = true;
-                break;
-            }
-        }
-
-        // final class... final function
-        if ((d.classDeclaration || d.functionDeclaration) && hasFinal)
-            pushIsFinal(true);
-
-        d.accept(this);
-
-        if ((d.classDeclaration || d.functionDeclaration) && hasFinal)
-            popIsFinal;
-    }
-
-    override void visit(const(FunctionCallExpression) exp)
-    {
-        // nested function are not virtual
-        pushNestedFunc(true);
-        exp.accept(this);
-        popNestedFunc();
-    }
-
-    override void visit(const(UnaryExpression) exp)
-    {
-        if (isInCtor)
-        // get function identifier for a call, only for this member (so no ident chain)
-        if (const IdentifierOrTemplateInstance iot = safeAccess(exp)
-            .functionCallExpression.unaryExpression.primaryExpression.identifierOrTemplateInstance)
-        {
-            const Token t = iot.identifier;
-            if (t != tok!"")
-            {
-                _ctorCalls[$-1] ~= t;
-            }
-        }
-        exp.accept(this);
-    }
-
-    override void visit(const(FunctionDeclaration) d)
-    {
-        if (isInClass() && !isInNestedFunc() && !isFinal() && !d.templateParameters)
-        {
-            bool virtualOnce;
-            bool notVirtualOnce;
-
-            const bool hasAttribs = d.attributes !is null;
-            const bool hasStatic = hasAttribs ? d.attributes.canFind!(a => a.attribute.type == tok!"static") : false;
-
-            // handle "private", "public"... for this declaration
-            if (d.attributes) foreach (attr; d.attributes.retro)
-            {
-                if (!hasStatic &&
-                   (attr.attribute == tok!"public" || attr.attribute == tok!"protected"))
-                {
-                    if (!isVirtual)
-                    {
-                        virtualOnce = true;
-                        break;
-                    }
-                }
-                else if (hasStatic || attr.attribute == tok!"private" || attr.attribute == tok!"package")
-                {
-                    if (isVirtual)
-                    {
-                        notVirtualOnce = true;
-                        break;
-                    }
-                }
-            }
-
-            if (!isVirtual && virtualOnce)
-                _virtualMethods[$-1] ~= d.name;
-            else if (isVirtual && !virtualOnce)
-                _virtualMethods[$-1] ~= d.name;
-
-        }
-        d.accept(this);
-    }
+	alias visit = BaseAnalyzerDmd.visit;
+	mixin AnalyzerInfo!"vcall_in_ctor";
+
+	extern (D) this(string fileName, bool skipTests = false)
+	{
+		super(fileName, skipTests);
+	}
+
+	override void visit(AST.ClassDeclaration d)
+	{
+		if (d.members)
+		{
+			foreach (s; *d.members)
+			{
+				if (s.isCtorDeclaration())
+					inClassCtor = true;
+
+				s.accept(this);
```

Comment:
This is not going to work. Consider this code:

```d
class D
{
       this(int a)
       {
              class T
              {
                      void bar() {}
                      void foobar () {  bar(); }
              }
       }
}

Reply:
I think you mean a code sequence of a constructor defining a class which also defines a constructor:
`class Foo
{
	this()
	{
		class Bar
		{
			this() { bar(); }
			private void bar();
		}

		foo(); // [warn]: %s
	}

	void foo() {}
}`
Good catch, will fix and add this example as a test.

Reply:
No, I meant my exact code. On my example, what I think your patch does is to output a warning when `bar` is called, which is wrong.

Also, the example you posted should also error if bar is not private.